interface: ignore interface with state 'unknown'

rajeshrah22 opened the pull request:
Fixes https://github.com/nmstate/nmstate/issues/2452

~~modify Interface::is_ignore to include InterfaceState::Unknown.~~
~~Log warn level in get_ignored_ifaces function.~~

Purge desired interfaces with state unknown in MergedInterfaces.
Log warning when doing so.

rajeshrah22 commented:
Problem: warning is printed twice.

rajeshrah22 commented:
Before:
```
$ echo '---
interfaces:
  - name: enp1s0
    type: ethernet
    state: unknown
    mtu: 9000' | sudo nmstatectl apply -
[2025-06-07T18:24:23Z INFO  nmstatectl] Nmstate version: 2.2.45
[2025-06-07T18:24:23Z INFO  nmstate::query_apply::net_state] Created checkpoint /org/freedesktop/NetworkManager/Checkpoint/4
[2025-06-07T18:24:23Z INFO  nmstate::nm::query_apply::profile] Modifying connection 28a5d59f-78b3-37f0-83d7-a01bb93661f4: enp1s0/802-3-ethernet
[2025-06-07T18:24:23Z INFO  nmstate::query_apply::net_state] Destroyed checkpoint /org/freedesktop/NetworkManager/Checkpoint/4
route-rules: {}
routes: {}
interfaces:
- name: enp1s0
  type: ethernet
  state: unknown
  mtu: 9000
ovn: {}
```

after:
```
t$ echo '---
interfaces:
  - name: enp1s0
    type: ethernet
    state: unknown
    mtu: 9000' | sudo ./target/release/nmstatectl apply -
[2025-06-07T18:25:23Z INFO  nmstatectl] Nmstate version: 2.2.46
[2025-06-07T18:25:23Z WARN  nmstate::ifaces::inter_ifaces] Interface enp1s0 has state Unknown, it will be ignored
[2025-06-07T18:25:23Z INFO  nmstate::ifaces::inter_ifaces] Ignoring interface enp1s0 type ethernet
[2025-06-07T18:25:23Z INFO  nmstate::query_apply::net_state] Created checkpoint /org/freedesktop/NetworkManager/Checkpoint/5
[2025-06-07T18:25:23Z INFO  nmstate::query_apply::net_state] Destroyed checkpoint /org/freedesktop/NetworkManager/Checkpoint/5
[2025-06-07T18:25:23Z WARN  nmstate::ifaces::inter_ifaces] Interface enp1s0 has state Unknown, it will be ignored
[2025-06-07T18:25:23Z INFO  nmstate::ifaces::inter_ifaces] Ignoring interface enp1s0 type ethernet
route-rules: {}
routes: {}
interfaces: []
ovn: {}
```

cathay4t commented on rust/src/lib/iface.rs:
```diff
@@ -566,6 +566,7 @@ impl Interface {
     /// Whether interface is marked as ignore.
     pub fn is_ignore(&self) -> bool {
         self.base_iface().state == InterfaceState::Ignore
+            || self.base_iface().state == InterfaceState::Unknown
```
The `state: unknown` is different from `state: ignore`:
 * `ignore` means nmstate should never touch it.
 * `unknown` means nmstate can touch it when required.

cathay4t reviewed (changes requested):
The warning and purging `state: unknown` interface should done at `MergedInterfaces`.

rajeshrah22 committed:
interface: Ignore interface with state 'unknown'

- add function to purge interfaces in desired interfaces with state
  InterfaceState::Unknown.
- add unit and intergration tests.
- add InterfaceState.UNKOWN to libnmstate/schema.py

Resolves: https://issues.redhat.com/browse/RHEL-96764

Signed-off-by: Rahul Rajesh <rajeshrah22@gmail.com>

rajeshrah22 commented:
Still the problem of twice printing log message is not solved.
I put all the logic in `MergedInterfaces::new`, which calls a function `resolve_unknown_state_ifaces`.

cathay4t commented on rust/src/lib/ifaces/inter_ifaces.rs:
```diff
@@ -355,6 +355,19 @@ impl Interfaces {
         Ok(())
     }
 
+    fn resolve_unknown_state_ifaces(&mut self) {
```
It is not resolve. And we should not change `state: unknown` to `state: ignore`.

We should **remove** these interfaces with `state: unknown`, so it still could be marked as changed when required.

If we change it to `state: ignore`, nmstate cannot touch it anymore.

rajeshrah22 replied:
So, removed them from the list of desired interfaces? And what do you mean "marked as changed"?

rajeshrah22 replied:
So, why might nmstate want to touch the interface with `state: unknown` again?

cathay4t replied:
This YAML will make eth1 as changed.

```yaml
---
interfaces:
- name: eth1
  state: unknown
- name: bond0
  type: bond
  state: up
  link-aggregation:
    mode: balance-rr
    port:
    - eth2
    - eth1
```

liangwen12year replied:
When a interface is marked as changed, the function `iface.post_inter_ifaces_process()` will be called.

cathay4t reviewed (changes requested):
Use `retain()` instead of generate a Vec of unknown interface.

rajeshrah22 commented:
@cathay4t I have used retain to remove unknown state interfaces from `kernel_ifaces` and `user_ifaces`. Should I also remove the corresponding interfaces in `insert_order`?

I am not fully sure how `insert_order` is used; however, it makes sense to keep the elements in `insert_order` consistent with what is available in the kernel and user interfaces.

cathay4t commented:
> @cathay4t I have used retain to remove unknown state interfaces from `kernel_ifaces` and `user_ifaces`. Should I also remove the corresponding interfaces in `insert_order`?
> 
> I am not fully sure how `insert_order` is used; however, it makes sense to keep the elements in `insert_order` consistent with what is available in the kernel and user interfaces.

The insert order to used for re-ordering interfaces base on their `controler-port-parent` relationship `set_ifaces_up_priority()`. `Vec` also has `retrain` function, please purge it also.

liangwen12year commented on rust/src/lib/ifaces/inter_ifaces.rs:
```diff
@@ -356,6 +356,24 @@ impl Interfaces {
+            iface.base_iface().state != InterfaceState::Unknown
+        });
+
+        self.insert_order.retain(|(iface_name, iface_type)| {
```
Can you add the following comments?
 
```
// Clean up insert_order Vec by removing entries for interfaces that were purged.
// The `insert_order` is used for re-ordering interfaces based on their controller-port-parent
// relationship in `set_ifaces_up_priority()`. We must maintain consistency between 
// `insert_order` and the HashMap collections to ensure proper interface dependency.
```

rajeshrah22 commented:
Thanks @cathay4t @liangwen12year !
